Inmutable objects in backend and fixes in performance tests

lucasponce opened the pull request:
@jotak, this is related our discussion on friday.
This approach was followed in Alerting 2.x, basically the idea is to separate the object used to store in the backend vs the object used in the Java API, and make it inmutable.
Also, this PR fixes performance tests and makes them mandatory in the profile.

jmazzitelli commented on hawkular-inventory-parent/hawkular-inventory-service/src/main/java/org/hawkular/inventory/service/ispn/IspnResource.java:
```diff
@@ -0,0 +1,97 @@
+
+    public Resource getResource() {
+        if (resource == null) {
+            return resource;
```
why 'return resource'? Shouldn't this be return null (yes, I realize resource is null because it wouldn't get here unless "if resource == null" is true, but still, seem odd and confusion). Should be "return null" just to be clear what it is that you intend to happen here (that is, return null when resource isn't set).

lucasponce replied:
Fixed.

jotak commented on hawkular-inventory-parent/hawkular-inventory-service/src/main/java/org/hawkular/inventory/service/InventoryServiceIspn.java:
```diff
@@ -149,7 +151,8 @@ public void deleteResourceType(String typeId) {
         }
-        return Optional.ofNullable((Resource) resource.get(id));
+        IspnResource ispnResource = (IspnResource) resource.get(id);
+        return Optional.ofNullable(ispnResource == null ? null : ispnResource.getResource());
```
not important, but that could be "return Optional.ofNullable(ispnResource).map(IspnResource::getResource)"

lucasponce replied:
+1

jotak commented on hawkular-inventory-parent/hawkular-inventory-service/src/main/java/org/hawkular/inventory/service/InventoryServiceIspn.java:
```diff
@@ -209,7 +218,8 @@ public void deleteResourceType(String typeId) {
         }
-        return Optional.ofNullable((ResourceType) resourceType.get(typeId));
+        IspnResourceType ispnResourceType = (IspnResourceType) resourceType.get(typeId);
+        return Optional.ofNullable(ispnResourceType == null ? null : ispnResourceType.getResourceType());
```
same here

lucasponce replied:
+1

jotak commented on hawkular-inventory-parent/hawkular-inventory-service/src/main/java/org/hawkular/inventory/service/InventoryServiceIspn.java:
```diff
@@ -279,17 +291,20 @@ public boolean isRunning() {
             return null;
         }
-        return (ResourceType) resourceType.get(typeId);
+        return ((IspnResourceType) resourceType.get(typeId)).getResourceType();
```
return value here can be null if the type doesn't exist, so it's missing a null-check

lucasponce replied:
+1 good catch, I missed this one

jotak commented on hawkular-inventory-parent/hawkular-inventory-service/src/main/java/org/hawkular/inventory/service/ispn/IspnResource.java:
```diff
@@ -0,0 +1,97 @@
+    @Field(store = Store.YES, analyze = Analyze.NO, indexNullAs = Field.DEFAULT_NULL_TOKEN)
+    private final String parentId;
+
+    private final Resource resource;
```
I'm not sure if that's ok. If the whole stateful "Resource" object is serialized with IspnResource, then we may have the same side-effects that I hit before, with lazy-loaded "children" being serialized as well here? The mutability of "resource" is going to leak to IspnResource.

jotak replied:
Oh, no because it's cloned, I understand.
But i'm still in favor of Immutable "Resource" :)

jotak replied:
Thinking again about it... it's cloned at construction, but then it's provided through getter without re-cloning, so it can still be mutated. I think it's a problem.

lucasponce commented:
I have addressed the comments of the PR.
I think this approach is interesting, as it separates API vs backend objects, and not bringing the hibernate-search dependencies to the Java API users (mostly cmdgw and listeners, as agent will serialize).
About the discussion Resource mutable vs several Resource classes inmutable depending if a resource brings children or type, is still being discussed in the #148 I guess this one could get in, and in the previous PR we can still discussing.

jotak commented:
Agree, I didn't mean to block this PR
